### PR TITLE
CLOUDSTACK-10073: KVM host RAM overprovisioning

### PR DESCRIPTION
**Pull request by bwsw**
The PR introduces new cloudstack agent KVM property
host.overcommit.mem.mb

which allows increasing KVM host RAM more than physical RAM exists. It is necessary to utilize KSM and ZSwap features which extend RAM with deduplication and compression.

### CODE REVIEW

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by bwsw** [comment]
@borisstoyanov I added the code with analogy of _dom0minMem parameter, It seems that there are no tests for it also and I don't feel comfortable enough with ACS background to determine the right place to implement units for the feature.

**Comment by borisstoyanov** [comment]
@blueorangutan test

File: agent/conf/agent.properties [hunk 1]
```diff
@@ -156,3 +156,7 @@ hypervisor.type=kvm
 #router.aggregation.command.each.timeout=600
 #timeout value for aggregation commands send to virtual router
 #
+# host.overcommit.mem.mb = 0
+# allows to increase amount of ram available on host virtually to utilize Zswap, KSM features
+# and modern fast SSD/3D XPoint devices
```
Comment (wido) [inline agent/conf/agent.properties]: You might want to add that it adds these X amount of memory to the memory it reports to the Management Server

File: plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java [hunk 2]
```diff
@@ -523,14 +523,7 @@ public Date startOfNextDay(Date dt) {
...
     public Date startOfNextDay() {
-        Calendar c = Calendar.getInstance();
-        c.setTime(new Date());
-        c.add(Calendar.DATE, 1);
-        c.set(Calendar.HOUR, 0);
-        c.set(Calendar.MINUTE, 0);
-        c.set(Calendar.SECOND, 0);
-        c.set(Calendar.MILLISECOND, 0);
-        return c.getTime();
+        return startOfNextDay(new Date());
```
Comment (wido) [inline plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java]: I wouldn't fix this in the same PR. Probably a separate PR for this fix would be better. It's not related to this change.

Comment (bwsw) [reply plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java]: @wido As today our Java guys found bug in Quota plugin related to the code above I ask them to fix it and we send a different PR for it.

File: agent/conf/agent.properties (hunk 1 above)
Comment (bwsw) [reply agent/conf/agent.properties]: @wido fixed, take a look.

File: plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java (hunk 2 above)
Comment (bwsw) [reply plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java]: @wido unrelated code is removed from PR.

**Comment by wido** [comment]
Looks good! Could you create a JIRA ticket and update the title of this PR?

In addition squash all commits and add the JIRA ticket to the subject/commit as well.

**Comment by bwsw** [comment]
@wido I tried to follow all the recommendations. Take a look please.

**Comment by wido** [comment]
LGTM based on the code

Interesting idea using 3D-Xpoint SSDs as SWAP :)

**Comment by bwsw** [comment]
@borisstoyanov 

```
sshClient: DEBUG: {Cmd: ssh -i ~/.ssh/id_rsa.cloud -ostricthostkeychecking=no -oUserKnownHostsFile=/dev/null -p 3922 169.254.3.66 /usr/local/cloud/systemvm/ssvm-check.sh |grep -e ERROR -e WARNING -e FAIL via Host: 10.2.2.67} {returns: [u'WARNING: cannot ping DNS server', u'Tests Complete. Look for ERROR or WARNING above.']}
paramiko.transport: DEBUG: EOF in transport thread
test_07_reboot_ssvm (tests.smoke.test_ssvm.TestSSVMs): DEBUG: SSVM script output: [u'WARNING: cannot ping DNS server', u'Tests Complete. Look for ERROR or WARNING above.']
test_07_reboot_ssvm (tests.smoke.test_ssvm.TestSSVMs): CRITICAL: FAILED: test_07_reboot_ssvm: ['Traceback (most recent call last):\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 369, in run\n    testMethod()\n', '  File "/marvin/tests/smoke/test_ssvm.py", line 929, in test_07_reboot_ssvm\n    self.test_03_ssvm_internals()\n', '  File "/marvin/tests/smoke/test_ssvm.py", line 449, in test_03_ssvm_internals\n    "Check for warnings in tests"\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 553, in assertEqual\n    assertion_func(first, second, msg=msg)\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 546, in _baseAssertEqual\n    raise self.failureException(msg)\n', 'AssertionError: Check for warnings in tests\n']
```

In test log before failure. I don't know if it's relevant to the case, because If it affects it should affect on VM start, beacause external specification doesn't change at all and default value for new parameter is 0, so it doesn't affect on actual test results.

**Comment by borisstoyanov** [comment]
How about the load balancer @bwsw ?

**Comment by wido** [comment]
I find it hard to believe a test failure is related to these changes. It doesn't change anything to the default behavior of the Agent. Still, I could be wrong.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
LGTM, merging

**Pull request merged**